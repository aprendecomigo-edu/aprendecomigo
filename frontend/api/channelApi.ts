import apiClient from './apiClient';

export interface User {
  id: number;
  username: string;
  email: string;
  first_name: string;
  last_name: string;
}

export interface Message {
  id: number;
  channel: number;
  sender: User;
  content: string;
  timestamp: string;
  file?: string;
  reactions: any[];
  attachments: any[];
}

export interface Channel {
  id: number;
  name: string;
  is_direct: boolean;
  participants: User[];
  online: User[];
  created_at: string;
  updated_at: string;
  last_message: Message | null;
}

/**
 * Fetch channels for the current user
 */
export const fetchChannels = async (): Promise<Channel[]> => {
  try {
    const response = await apiClient.get('/channels/');
    if (__DEV__) {
      console.log('Backend channels response:', response.data);
    }
    // Handle paginated response from Django REST Framework
    if (response.data && Array.isArray(response.data.results)) {
      return response.data.results;
    }
    // Handle direct array response (fallback)
    if (Array.isArray(response.data)) {
      return response.data;
    }
    if (__DEV__) {
      console.warn('Unexpected API response format:', response.data);
    }
    return [];
  } catch (error) {
    console.error('Error fetching channels:', error);
    return [];
  }
};

/**
 * Fetch messages for a specific channel
 */
export const fetchMessages = async (channelId: number): Promise<Message[]> => {
  try {
    const response = await apiClient.get(`/channels/${channelId}/messages/`);
    return response.data;
  } catch (error) {
    console.error('Error fetching messages:', error);
    return [];
  }
};

/**
 * Send a message to a channel
 */
export const sendMessage = async (
  channelId: number,
  content: string,
  file?: File,
): Promise<Message> => {
  try {
    const formData = new FormData();
    formData.append('content', content);
    if (file) {
      formData.append('file', file);
    }

    const response = await apiClient.post(`/channels/${channelId}/send_message/`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    return response.data;
  } catch (error) {
    console.error('Error sending message:', error);
    throw error;
  }
};

/**
 * Create a new channel
 */
export const createChannel = async (
  name: string,
  isDirect: boolean,
  participantIds: number[],
): Promise<Channel> => {
  try {
    const response = await apiClient.post('/channels/', {
      name,
      is_direct: isDirect,
      participant_ids: participantIds,
    });
    return response.data;
  } catch (error) {
    console.error('Error creating channel:', error);
    throw error;
  }
};

/**
 * Search users
 */
export const searchUsers = async (query: string): Promise<User[]> => {
  try {
    const response = await apiClient.get(`/users/?search=${encodeURIComponent(query)}`);
    if (__DEV__) {
      console.log('Backend users search response:', response.data);
    }
    // Handle paginated response from Django REST Framework
    if (response.data && Array.isArray(response.data.results)) {
      return response.data.results;
    }
    // Handle direct array response (fallback)
    if (Array.isArray(response.data)) {
      return response.data;
    }
    if (__DEV__) {
      console.warn('Unexpected users API response format:', response.data);
    }
    return [];
  } catch (error) {
    console.error('Error searching users:', error);
    return [];
  }
};
