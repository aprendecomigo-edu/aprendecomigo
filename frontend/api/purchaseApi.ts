/**
 * API client functions for purchase-related operations.
 *
 * Handles communication with the backend purchase APIs including
 * pricing plans, purchase initiation, student balance, and Stripe configuration.
 */

import apiClient from './apiClient';

import type {
  PricingPlan,
  PurchaseInitiationRequest,
  PurchaseInitiationResponse,
  StudentBalanceResponse,
  StripeConfig,
  PaginatedTransactionHistory,
  PaginatedPurchaseHistory,
  TransactionFilterOptions,
  PurchaseFilterOptions,
  TopUpPackage,
  RenewalRequest,
  RenewalResponse,
  QuickTopUpRequest,
  QuickTopUpResponse,
} from '@/types/purchase';

/**
 * Purchase API client with comprehensive error handling and type safety.
 */
export class PurchaseApiClient {
  /**
   * Fetch all active pricing plans.
   *
   * @returns Promise resolving to array of active pricing plans
   * @throws Error with descriptive message if request fails
   */
  static async getPricingPlans(): Promise<PricingPlan[]> {
    try {
      const response = await apiClient.get('/finances/pricing-plans/');

      if (!Array.isArray(response.data)) {
        throw new Error('Invalid response format: expected array of pricing plans');
      }

      return response.data.map((plan: any) => ({
        id: plan.id,
        name: plan.name,
        description: plan.description,
        plan_type: plan.plan_type,
        plan_type_display: plan.plan_type_display,
        hours_included: plan.hours_included,
        price_eur: plan.price_eur,
        price_per_hour: plan.price_per_hour,
        validity_days: plan.validity_days,
        is_active: plan.is_active,
        display_order: plan.display_order,
      }));
    } catch (error: any) {
      console.error('Error fetching pricing plans:', error);

      if (error.response?.status === 503) {
        throw new Error('Pricing service temporarily unavailable. Please try again later.');
      } else if (error.response?.status >= 500) {
        throw new Error('Server error occurred while loading pricing plans');
      } else if (error.code === 'NETWORK_ERROR' || error.message === 'Network Error') {
        throw new Error('Network connection error. Please check your internet connection.');
      } else {
        throw new Error('Failed to load pricing plans. Please try again.');
      }
    }
  }

  /**
   * Initiate a purchase for a specific pricing plan.
   *
   * @param request Purchase initiation request data
   * @returns Promise resolving to purchase initiation response
   * @throws Error with descriptive message if request fails
   */
  static async initiatePurchase(
    request: PurchaseInitiationRequest,
  ): Promise<PurchaseInitiationResponse> {
    try {
      const response = await apiClient.post('/finances/purchase/initiate/', request);

      return {
        success: response.data.success,
        client_secret: response.data.client_secret,
        transaction_id: response.data.transaction_id,
        payment_intent_id: response.data.payment_intent_id,
        plan_details: response.data.plan_details,
        message: response.data.message,
      };
    } catch (error: any) {
      console.error('Error initiating purchase:', error);

      // Handle validation errors with detailed field information
      if (error.response?.status === 400 && error.response.data) {
        const errorData = error.response.data;
        return {
          success: false,
          error_type: errorData.error_type || 'validation_error',
          message: errorData.message || 'Invalid request data',
          field_errors: errorData.field_errors || {},
        };
      }

      // Handle rate limiting
      if (error.response?.status === 429) {
        throw new Error('Too many purchase attempts. Please try again later.');
      }

      // Handle server errors
      if (error.response?.status >= 500) {
        throw new Error('Server error occurred during purchase initiation');
      }

      // Handle network errors
      if (error.code === 'NETWORK_ERROR' || error.message === 'Network Error') {
        throw new Error('Network connection error. Please check your internet connection.');
      }

      // Handle payment service errors
      if (error.response?.data?.error_type === 'api_connection_error') {
        throw new Error('Payment service temporarily unavailable. Please try again later.');
      }

      // Generic error fallback
      const errorMessage = error.response?.data?.message || 'Failed to initiate purchase';
      throw new Error(errorMessage);
    }
  }

  /**
   * Get student balance information.
   *
   * @param email Optional email parameter for admin access
   * @returns Promise resolving to student balance data
   * @throws Error with descriptive message if request fails
   */
  static async getStudentBalance(email?: string): Promise<StudentBalanceResponse> {
    try {
      const params = email ? { email } : {};
      const response = await apiClient.get('/finances/student-balance/', { params });

      return {
        student_info: response.data.student_info,
        balance_summary: response.data.balance_summary,
        package_status: response.data.package_status,
        upcoming_expirations: response.data.upcoming_expirations,
      };
    } catch (error: any) {
      console.error('Error fetching student balance:', error);

      if (error.response?.status === 404) {
        throw new Error('Student not found');
      } else if (error.response?.status === 403) {
        throw new Error('Access denied. Unable to retrieve balance information.');
      } else if (error.response?.status >= 500) {
        throw new Error('Server error occurred while loading balance information');
      } else if (error.code === 'NETWORK_ERROR' || error.message === 'Network Error') {
        throw new Error('Network connection error. Please check your internet connection.');
      } else {
        throw new Error('Failed to load balance information. Please try again.');
      }
    }
  }

  /**
   * Get Stripe configuration for frontend initialization.
   *
   * @returns Promise resolving to Stripe configuration
   * @throws Error with descriptive message if request fails
   */
  static async getStripeConfig(): Promise<StripeConfig> {
    try {
      const response = await apiClient.get('/finances/stripe/config/');

      return {
        public_key: response.data.public_key,
        success: response.data.success,
      };
    } catch (error: any) {
      console.error('Error fetching Stripe config:', error);

      if (error.response?.status === 503) {
        throw new Error('Payment service temporarily unavailable');
      } else if (error.response?.status >= 500) {
        throw new Error('Server error occurred while loading payment configuration');
      } else if (error.code === 'NETWORK_ERROR' || error.message === 'Network Error') {
        throw new Error('Network connection error. Please check your internet connection.');
      } else {
        throw new Error('Failed to load payment configuration. Please try again.');
      }
    }
  }

  /**
   * Check if a student can book a session of specified duration.
   *
   * @param durationHours Session duration in hours
   * @param email Optional email parameter for admin access
   * @returns Promise resolving to booking eligibility information
   * @throws Error with descriptive message if request fails
   */
  static async checkBookingEligibility(durationHours: number, email?: string): Promise<any> {
    try {
      const params: any = { duration_hours: durationHours.toString() };
      if (email) {
        params.email = email;
      }

      const response = await apiClient.get('/finances/student-balance/check-booking/', {
        params,
      });
      return response.data;
    } catch (error: any) {
      console.error('Error checking booking eligibility:', error);

      if (error.response?.status === 400) {
        throw new Error(error.response.data?.error || 'Invalid booking request');
      } else if (error.response?.status === 404) {
        throw new Error('Student not found');
      } else if (error.response?.status >= 500) {
        throw new Error('Server error occurred while checking booking eligibility');
      } else if (error.code === 'NETWORK_ERROR' || error.message === 'Network Error') {
        throw new Error('Network connection error. Please check your internet connection.');
      } else {
        throw new Error('Failed to check booking eligibility. Please try again.');
      }
    }
  }

  /**
   * Get student transaction history.
   *
   * @param options Query options for filtering and pagination
   * @returns Promise resolving to paginated transaction history
   * @throws Error with descriptive message if request fails
   */
  static async getTransactionHistory(
    options: TransactionFilterOptions & {
      page?: number;
      page_size?: number;
      email?: string;
    } = {},
  ): Promise<PaginatedTransactionHistory> {
    try {
      const response = await apiClient.get('/finances/student-balance/history/', {
        params: options,
      });
      return response.data;
    } catch (error: any) {
      console.error('Error fetching transaction history:', error);

      if (error.response?.status === 404) {
        throw new Error('Student not found');
      } else if (error.response?.status === 403) {
        throw new Error('Access denied. Unable to retrieve transaction history.');
      } else if (error.response?.status >= 500) {
        throw new Error('Server error occurred while loading transaction history');
      } else if (error.code === 'NETWORK_ERROR' || error.message === 'Network Error') {
        throw new Error('Network connection error. Please check your internet connection.');
      } else {
        throw new Error('Failed to load transaction history. Please try again.');
      }
    }
  }

  /**
   * Get detailed purchase history with consumption tracking.
   *
   * @param options Query options for filtering and pagination
   * @returns Promise resolving to detailed purchase history
   * @throws Error with descriptive message if request fails
   */
  static async getPurchaseHistory(
    options: PurchaseFilterOptions & {
      page?: number;
      page_size?: number;
      email?: string;
    } = {},
  ): Promise<PaginatedPurchaseHistory> {
    try {
      const params: any = { ...options };
      if (options.active_only !== undefined) {
        params.active_only = options.active_only.toString();
      }
      if (options.include_consumption !== undefined) {
        params.include_consumption = options.include_consumption.toString();
      }

      const response = await apiClient.get('/finances/student-balance/purchases/', {
        params,
      });
      return response.data;
    } catch (error: any) {
      console.error('Error fetching purchase history:', error);

      if (error.response?.status === 404) {
        throw new Error('Student not found');
      } else if (error.response?.status === 403) {
        throw new Error('Access denied. Unable to retrieve purchase history.');
      } else if (error.response?.status >= 500) {
        throw new Error('Server error occurred while loading purchase history');
      } else if (error.code === 'NETWORK_ERROR' || error.message === 'Network Error') {
        throw new Error('Network connection error. Please check your internet connection.');
      } else {
        throw new Error('Failed to load purchase history. Please try again.');
      }
    }
  }

  /**
   * Get student receipts.
   *
   * @param email Optional email parameter for admin access
   * @returns Promise resolving to array of receipts
   * @throws Error with descriptive message if request fails
   */
  static async getReceipts(email?: string): Promise<any[]> {
    try {
      const params = email ? { email } : {};
      const response = await apiClient.get('/finances/student-balance/receipts/', { params });
      return response.data;
    } catch (error: any) {
      console.error('Error fetching receipts:', error);
      throw new Error('Failed to load receipts. Please try again.');
    }
  }

  /**
   * Generate a receipt for a transaction.
   *
   * @param transactionId Transaction ID to generate receipt for
   * @param email Optional email parameter for admin access
   * @returns Promise resolving to receipt generation response
   * @throws Error with descriptive message if request fails
   */
  static async generateReceipt(transactionId: string, email?: string): Promise<any> {
    try {
      const data = email
        ? { transaction_id: transactionId, email }
        : { transaction_id: transactionId };
      const response = await apiClient.post('/finances/student-balance/receipts/generate/', data);
      return response.data;
    } catch (error: any) {
      console.error('Error generating receipt:', error);
      throw new Error('Failed to generate receipt. Please try again.');
    }
  }

  /**
   * Get student payment methods.
   *
   * @param email Optional email parameter for admin access
   * @returns Promise resolving to array of payment methods
   * @throws Error with descriptive message if request fails
   */
  static async getPaymentMethods(email?: string): Promise<any[]> {
    try {
      const params = email ? { email } : {};
      const response = await apiClient.get('/finances/student-balance/payment-methods/', {
        params,
      });
      return response.data;
    } catch (error: any) {
      console.error('Error fetching payment methods:', error);
      throw new Error('Failed to load payment methods. Please try again.');
    }
  }

  /**
   * Add a new payment method.
   *
   * @param paymentMethodId Stripe payment method ID
   * @param setAsDefault Whether to set as default payment method
   * @param email Optional email parameter for admin access
   * @returns Promise resolving to payment method addition response
   * @throws Error with descriptive message if request fails
   */
  static async addPaymentMethod(
    paymentMethodId: string,
    setAsDefault: boolean = false,
    email?: string,
  ): Promise<any> {
    try {
      const data = {
        payment_method_id: paymentMethodId,
        set_as_default: setAsDefault,
        ...(email && { email }),
      };
      const response = await apiClient.post('/finances/student-balance/payment-methods/', data);
      return response.data;
    } catch (error: any) {
      console.error('Error adding payment method:', error);
      throw new Error('Failed to add payment method. Please try again.');
    }
  }

  /**
   * Get usage analytics.
   *
   * @param timeRange Optional time range for analytics
   * @param email Optional email parameter for admin access
   * @returns Promise resolving to usage analytics
   * @throws Error with descriptive message if request fails
   */
  static async getUsageAnalytics(
    timeRange?: { start_date: string; end_date: string },
    email?: string,
  ): Promise<any> {
    try {
      const params: any = {};
      if (email) params.email = email;
      if (timeRange) {
        params.start_date = timeRange.start_date;
        params.end_date = timeRange.end_date;
      }

      const response = await apiClient.get('/finances/student-balance/analytics/usage/', {
        params,
      });
      return response.data;
    } catch (error: any) {
      console.error('Error fetching usage analytics:', error);
      throw new Error('Failed to load usage analytics. Please try again.');
    }
  }

  /**
   * Get available top-up packages for quick purchase.
   *
   * @param email Optional email parameter for admin access
   * @returns Promise resolving to array of top-up packages
   * @throws Error with descriptive message if request fails
   */
  static async getTopUpPackages(email?: string): Promise<TopUpPackage[]> {
    try {
      const params = email ? { email } : {};
      const response = await apiClient.get('/finances/student-balance/topup-packages/', { params });

      if (!Array.isArray(response.data)) {
        throw new Error('Invalid response format: expected array of top-up packages');
      }

      return response.data.map((pkg: any) => ({
        id: pkg.id,
        name: pkg.name,
        hours: pkg.hours,
        price_eur: pkg.price_eur,
        price_per_hour: pkg.price_per_hour,
        is_popular: pkg.is_popular,
        discount_percentage: pkg.discount_percentage,
        display_order: pkg.display_order,
      }));
    } catch (error: any) {
      console.error('Error fetching top-up packages:', error);

      if (error.response?.status === 503) {
        throw new Error('Top-up service temporarily unavailable. Please try again later.');
      } else if (error.response?.status >= 500) {
        throw new Error('Server error occurred while loading top-up packages');
      } else if (error.code === 'NETWORK_ERROR' || error.message === 'Network Error') {
        throw new Error('Network connection error. Please check your internet connection.');
      } else {
        throw new Error('Failed to load top-up packages. Please try again.');
      }
    }
  }

  /**
   * Renew expired subscription with saved payment method.
   *
   * @param request Renewal request data
   * @param email Optional email parameter for admin access
   * @returns Promise resolving to renewal response
   * @throws Error with descriptive message if request fails
   */
  static async renewSubscription(
    request: RenewalRequest,
    email?: string,
  ): Promise<RenewalResponse> {
    try {
      const data = email ? { ...request, email } : request;
      const response = await apiClient.post('/finances/student-balance/renew-subscription/', data);

      return {
        success: response.data.success,
        transaction_id: response.data.transaction_id,
        payment_intent_id: response.data.payment_intent_id,
        renewal_details: response.data.renewal_details,
        message: response.data.message,
        client_secret: response.data.client_secret,
      };
    } catch (error: any) {
      console.error('Error renewing subscription:', error);

      // Handle validation errors with detailed field information
      if (error.response?.status === 400 && error.response.data) {
        const errorData = error.response.data;
        return {
          success: false,
          error_type: errorData.error_type || 'validation_error',
          message: errorData.message || 'Invalid renewal request',
          field_errors: errorData.field_errors || {},
        };
      }

      // Handle payment errors
      if (error.response?.status === 402) {
        throw new Error('Payment could not be processed. Please check your payment method.');
      }

      // Handle rate limiting
      if (error.response?.status === 429) {
        throw new Error('Too many renewal attempts. Please try again later.');
      }

      // Handle server errors
      if (error.response?.status >= 500) {
        throw new Error('Server error occurred during subscription renewal');
      }

      // Handle network errors
      if (error.code === 'NETWORK_ERROR' || error.message === 'Network Error') {
        throw new Error('Network connection error. Please check your internet connection.');
      }

      // Handle payment service errors
      if (error.response?.data?.error_type === 'api_connection_error') {
        throw new Error('Payment service temporarily unavailable. Please try again later.');
      }

      // Generic error fallback
      const errorMessage = error.response?.data?.message || 'Failed to renew subscription';
      throw new Error(errorMessage);
    }
  }

  /**
   * Purchase additional hours quickly with saved payment method.
   *
   * @param request Quick top-up request data
   * @param email Optional email parameter for admin access
   * @returns Promise resolving to top-up response
   * @throws Error with descriptive message if request fails
   */
  static async quickTopUp(request: QuickTopUpRequest, email?: string): Promise<QuickTopUpResponse> {
    try {
      const data = email ? { ...request, email } : request;
      const response = await apiClient.post('/finances/student-balance/quick-topup/', data);

      return {
        success: response.data.success,
        transaction_id: response.data.transaction_id,
        payment_intent_id: response.data.payment_intent_id,
        package_details: response.data.package_details,
        message: response.data.message,
        client_secret: response.data.client_secret,
      };
    } catch (error: any) {
      console.error('Error processing quick top-up:', error);

      // Handle validation errors with detailed field information
      if (error.response?.status === 400 && error.response.data) {
        const errorData = error.response.data;
        return {
          success: false,
          error_type: errorData.error_type || 'validation_error',
          message: errorData.message || 'Invalid top-up request',
          field_errors: errorData.field_errors || {},
        };
      }

      // Handle payment errors
      if (error.response?.status === 402) {
        throw new Error('Payment could not be processed. Please check your payment method.');
      }

      // Handle rate limiting
      if (error.response?.status === 429) {
        throw new Error('Too many purchase attempts. Please try again later.');
      }

      // Handle server errors
      if (error.response?.status >= 500) {
        throw new Error('Server error occurred during top-up purchase');
      }

      // Handle network errors
      if (error.code === 'NETWORK_ERROR' || error.message === 'Network Error') {
        throw new Error('Network connection error. Please check your internet connection.');
      }

      // Handle payment service errors
      if (error.response?.data?.error_type === 'api_connection_error') {
        throw new Error('Payment service temporarily unavailable. Please try again later.');
      }

      // Generic error fallback
      const errorMessage = error.response?.data?.message || 'Failed to process top-up purchase';
      throw new Error(errorMessage);
    }
  }
}

// Convenience exports for direct function access
export const {
  getPricingPlans,
  initiatePurchase,
  getStudentBalance,
  getStripeConfig,
  checkBookingEligibility,
  getTransactionHistory,
  getPurchaseHistory,
  getReceipts,
  generateReceipt,
  getPaymentMethods,
  addPaymentMethod,
  getUsageAnalytics,
  getTopUpPackages,
  renewSubscription,
  quickTopUp,
} = PurchaseApiClient;
