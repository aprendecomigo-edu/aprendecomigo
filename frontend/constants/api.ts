/**
 * API URL configuration for different environments.
 *
 * This file automatically resolves to the appropriate platform-specific implementation:
 * - api.web.ts for web platforms
 * - api.native.ts for iOS/Android platforms
 */

import { ENVIRONMENT } from './env';

// Fallback configuration (should be overridden by platform-specific files)
const config = {
  development: {
    API_URL: 'http://localhost:8000/api',
  },
  staging: {
    API_URL: 'https://aprendecomigo.eu.pythonanywhere.com/api',
  },
  production: {
    API_URL: 'https://api.aprendecomigo.com/api',
  },
};

// Select the appropriate configuration
const activeConfig = config[ENVIRONMENT] || config.development;

// Export the API URL (will be overridden by platform-specific files)
export const API_URL = activeConfig.API_URL;

// Debug information
if (__DEV__) {
  console.log(`API URL (Fallback): ${API_URL}`);
}
