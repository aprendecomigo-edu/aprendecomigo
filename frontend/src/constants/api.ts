/**
 * API constants for different environments
 */
import { DEV_API_URL, PROD_API_URL } from '@env';

// Current API URL based on environment
export const API_URL = __DEV__ ? DEV_API_URL : PROD_API_URL;
