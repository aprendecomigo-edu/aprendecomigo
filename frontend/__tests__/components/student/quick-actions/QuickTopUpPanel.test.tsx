/**
 * QuickTopUpPanel Component Tests - Basic Implementation
 *
 * Tests the component behavior using toJSON snapshots and basic structure validation
 * since text queries don't work reliably with Gluestack UI mocking.
 */

import React from 'react';
import { render, waitFor } from '@testing-library/react-native';
import { QuickTopUpPanel } from '@/components/student/quick-actions/QuickTopUpPanel';
import { usePaymentMethods } from '@/hooks/usePaymentMethods';
import { useStudentBalance } from '@/hooks/useStudentBalance';
import { PurchaseApiClient } from '@/api/purchaseApi';

// Mock the hooks
jest.mock('@/hooks/usePaymentMethods');
jest.mock('@/hooks/useStudentBalance');
jest.mock('@/api/purchaseApi');

// Mock UI components
jest.mock('@/components/ui/toast', () => ({
  useToast: () => ({
    show: jest.fn(),
  }),
}));

// Mock dependency context
jest.mock('@/services/di/context', () => ({
  useDependencies: () => ({
    paymentService: {
      processQuickTopUp: jest.fn().mockResolvedValue({
        packageId: '5h-package',
        paymentMethodId: null,
      }),
    },
  }),
}));

const mockUsePaymentMethods = usePaymentMethods as jest.MockedFunction<typeof usePaymentMethods>;
const mockUseStudentBalance = useStudentBalance as jest.MockedFunction<typeof useStudentBalance>;
const mockPurchaseApiClient = PurchaseApiClient as jest.Mocked<typeof PurchaseApiClient>;

// Test data
const mockPackages = [
  {
    id: '5h-package',
    name: '5 Hours',
    hours: 5,
    price_eur: 25.0,
    price_per_hour: 5.0,
    discount_percentage: 0,
    is_popular: false,
    display_order: 1,
  },
  {
    id: '10h-package',
    name: '10 Hours',
    hours: 10,
    price_eur: 45.0,
    price_per_hour: 4.5,
    discount_percentage: 10,
    is_popular: true,
    display_order: 2,
  },
];

const mockPaymentMethods = [
  {
    id: 'pm_test123',
    card: {
      brand: 'visa',
      last4: '4242',
      exp_month: 12,
      exp_year: 2025,
    },
    is_default: true,
  },
];

describe('QuickTopUpPanel Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    // Setup default hook returns
    mockUsePaymentMethods.mockReturnValue({
      paymentMethods: mockPaymentMethods,
      loading: false,
      error: null,
      removing: false,
      settingDefault: false,
      operationError: null,
      refreshPaymentMethods: jest.fn(),
      removePaymentMethod: jest.fn(),
      setDefaultPaymentMethod: jest.fn(),
      clearErrors: jest.fn(),
      hasPaymentMethods: true,
    });

    mockUseStudentBalance.mockReturnValue({
      balance: null,
      loading: false,
      error: null,
      refetch: jest.fn(),
    });

    // Setup API mock
    mockPurchaseApiClient.getTopUpPackages.mockResolvedValue(mockPackages);
    mockPurchaseApiClient.quickTopUp.mockResolvedValue({
      success: true,
      message: 'Purchase successful',
      balance: { hours: 15 },
    });
  });

  describe('Component Rendering', () => {
    it('should render without crashing', () => {
      const { toJSON } = render(<QuickTopUpPanel />);
      expect(toJSON()).toBeTruthy();
    });

    it('should render loading state initially', () => {
      const { toJSON } = render(<QuickTopUpPanel />);
      
      // Component should render some structure
      expect(toJSON()).not.toBeNull();
      
      // Check that it contains loading state elements
      const rendered = toJSON();
      expect(rendered).toBeTruthy();
    });

    it('should have consistent structure', () => {
      const { toJSON } = render(<QuickTopUpPanel />);
      
      // Snapshot testing to ensure structure doesn't break
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render with different props', () => {
      const { toJSON } = render(
        <QuickTopUpPanel 
          email="test@example.com" 
          isModal={true}
          onTopUpSuccess={jest.fn()}
          onTopUpError={jest.fn()}
          onClose={jest.fn()}
        />
      );
      
      expect(toJSON()).toBeTruthy();
    });
  });

  describe('Hook Integration', () => {
    it('should call usePaymentMethods hook', () => {
      render(<QuickTopUpPanel />);
      expect(mockUsePaymentMethods).toHaveBeenCalled();
    });

    it('should pass email to usePaymentMethods', () => {
      render(<QuickTopUpPanel email="test@example.com" />);
      expect(mockUsePaymentMethods).toHaveBeenCalledWith('test@example.com');
    });

    it('should call useStudentBalance hook', () => {
      render(<QuickTopUpPanel />);
      expect(mockUseStudentBalance).toHaveBeenCalled();
    });

    it('should pass email to useStudentBalance', () => {
      render(<QuickTopUpPanel email="test@example.com" />);
      expect(mockUseStudentBalance).toHaveBeenCalledWith('test@example.com');
    });
  });

  describe('API Integration', () => {
    it('should call getTopUpPackages on mount', async () => {
      render(<QuickTopUpPanel />);
      
      await waitFor(() => {
        expect(mockPurchaseApiClient.getTopUpPackages).toHaveBeenCalled();
      });
    });

    it('should pass email to API call', async () => {
      render(<QuickTopUpPanel email="test@example.com" />);
      
      await waitFor(() => {
        expect(mockPurchaseApiClient.getTopUpPackages).toHaveBeenCalledWith('test@example.com');
      });
    });

    it('should handle API calls without email', async () => {
      render(<QuickTopUpPanel />);
      
      await waitFor(() => {
        expect(mockPurchaseApiClient.getTopUpPackages).toHaveBeenCalledWith(undefined);
      });
    });
  });

  describe('Error Handling', () => {
    it('should handle API errors gracefully', async () => {
      mockPurchaseApiClient.getTopUpPackages.mockRejectedValue(new Error('Network error'));

      const { toJSON } = render(<QuickTopUpPanel />);
      
      // Should still render without crashing
      await waitFor(() => {
        expect(toJSON()).toBeTruthy();
      });
    });

    it('should handle empty packages response', async () => {
      mockPurchaseApiClient.getTopUpPackages.mockResolvedValue([]);

      const { toJSON } = render(<QuickTopUpPanel />);
      
      await waitFor(() => {
        expect(toJSON()).toBeTruthy();
      });
    });
  });

  describe('Payment Method States', () => {
    it('should handle no payment methods', () => {
      mockUsePaymentMethods.mockReturnValue({
        paymentMethods: [],
        loading: false,
        error: null,
        removing: false,
        settingDefault: false,
        operationError: null,
        refreshPaymentMethods: jest.fn(),
        removePaymentMethod: jest.fn(),
        setDefaultPaymentMethod: jest.fn(),
        clearErrors: jest.fn(),
        hasPaymentMethods: false,
      });

      const { toJSON } = render(<QuickTopUpPanel />);
      expect(toJSON()).toBeTruthy();
    });

    it('should handle loading payment methods', () => {
      mockUsePaymentMethods.mockReturnValue({
        paymentMethods: [],
        loading: true,
        error: null,
        removing: false,
        settingDefault: false,
        operationError: null,
        refreshPaymentMethods: jest.fn(),
        removePaymentMethod: jest.fn(),
        setDefaultPaymentMethod: jest.fn(),
        clearErrors: jest.fn(),
        hasPaymentMethods: false,
      });

      const { toJSON } = render(<QuickTopUpPanel />);
      expect(toJSON()).toBeTruthy();
    });

    it('should handle payment method errors', () => {
      mockUsePaymentMethods.mockReturnValue({
        paymentMethods: [],
        loading: false,
        error: 'Failed to load payment methods',
        removing: false,
        settingDefault: false,
        operationError: null,
        refreshPaymentMethods: jest.fn(),
        removePaymentMethod: jest.fn(),
        setDefaultPaymentMethod: jest.fn(),
        clearErrors: jest.fn(),
        hasPaymentMethods: false,
      });

      const { toJSON } = render(<QuickTopUpPanel />);
      expect(toJSON()).toBeTruthy();
    });
  });

  describe('Balance Service Integration', () => {
    it('should handle balance loading state', () => {
      mockUseStudentBalance.mockReturnValue({
        balance: null,
        loading: true,
        error: null,
        refetch: jest.fn(),
      });

      const { toJSON } = render(<QuickTopUpPanel />);
      expect(toJSON()).toBeTruthy();
    });

    it('should handle balance errors', () => {
      mockUseStudentBalance.mockReturnValue({
        balance: null,
        loading: false,
        error: 'Failed to load balance',
        refetch: jest.fn(),
      });

      const { toJSON } = render(<QuickTopUpPanel />);
      expect(toJSON()).toBeTruthy();
    });

    it('should handle successful balance data', () => {
      mockUseStudentBalance.mockReturnValue({
        balance: { hours: 10 },
        loading: false,
        error: null,
        refetch: jest.fn(),
      });

      const { toJSON } = render(<QuickTopUpPanel />);
      expect(toJSON()).toBeTruthy();
    });
  });

  describe('Performance', () => {
    it('should render efficiently', () => {
      const start = performance.now();
      render(<QuickTopUpPanel />);
      const end = performance.now();

      expect(end - start).toBeLessThan(100);
    });

    it('should handle re-renders efficiently', () => {
      const { rerender } = render(<QuickTopUpPanel />);

      const start = performance.now();
      rerender(<QuickTopUpPanel email="test@example.com" />);
      const end = performance.now();

      expect(end - start).toBeLessThan(50);
    });
  });

  describe('Prop Validation', () => {
    it('should accept all optional props', () => {
      const onTopUpSuccess = jest.fn();
      const onTopUpError = jest.fn();
      const onClose = jest.fn();

      const { toJSON } = render(
        <QuickTopUpPanel
          email="test@example.com"
          onTopUpSuccess={onTopUpSuccess}
          onTopUpError={onTopUpError}
          isModal={true}
          onClose={onClose}
        />
      );

      expect(toJSON()).toBeTruthy();
    });

    it('should work without any props', () => {
      const { toJSON } = render(<QuickTopUpPanel />);
      expect(toJSON()).toBeTruthy();
    });
  });

  describe('Component State Management', () => {
    it('should handle component lifecycle correctly', async () => {
      const { unmount } = render(<QuickTopUpPanel />);
      
      // Should not throw when unmounting
      expect(() => unmount()).not.toThrow();
    });

    it('should handle prop changes', () => {
      const { rerender } = render(<QuickTopUpPanel />);
      
      // Should not throw when props change
      expect(() => {
        rerender(<QuickTopUpPanel email="test@example.com" />);
      }).not.toThrow();
    });
  });
});