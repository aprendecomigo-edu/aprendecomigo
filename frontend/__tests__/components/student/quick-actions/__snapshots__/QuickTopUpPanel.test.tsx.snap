// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuickTopUpPanel Component Component Rendering should have consistent structure 1`] = `
<div
  className="gluestack-card"
  data-testid="Card"
  ref={null}
>
  <div
    className="gluestack-vstack"
    data-testid="VStack"
    ref={null}
    space="lg"
  >
    <div
      className="gluestack-spinner"
      data-testid="Spinner"
      ref={null}
      size="large"
    />
    <div
      className="gluestack-text"
      data-testid="Text"
      ref={null}
    >
      Loading packages...
    </div>
  </div>
</div>
`;
